README.md
Barcode Bundle
==============

Generate barcodes as html, image or text.

Fork of Folke Ashberg's PHP Barcode v0.4 [http://www.ashberg.de/php-barcode/]


For usage and examples see: Resources/doc/index.rst


Requirements
------------

PHP 5.3.2


Optional
--------

Twig
GD Lib
genbarcode

If you want to generate not only EAN-12/EAN-13/ISBN-Codes you have to install
genbarcode, a small unix-commandline tool which uses GNU-Barcode.
genbarcode is available http://www.ashberg.de/php-barcode , read genbarcodes
README for installation.
If you have installed genbarcode not to /usr/bin call set $barcode->setGenbarcodeLocation('/usr/local/bin/genbarcode')


Installation
------------

Add HackzillaBarcodeBundle in your composer.json:

```js
{
    "require": {
        "hackzilla/barcode-bundle": "*"
    }
}
```

Now tell composer to download the bundle by running the command:

``` bash
$ php composer.phar update hackzilla/barcode-bundle
```

Composer will install the bundle to your project's `vendor/hackzilla` directory.

### Step 2: Enable the bundle

Enable the bundle in the kernel:

``` php
<?php
// app/AppKernel.php

public function registerBundles()
{
    $bundles = array(
        // ...
        new Hackzilla\BarcodeBundle\HackzillaBarcodeBundle(),
    );
}
```


CHANGELOG
Version 0.4, 2011-08-15, Folke Ashberg <folke@ashberg.de>
 * Bugfix: Correct Shell-Escaping calling genbarcode
 * Use of preg functions instead of ereg
 * Shipping Free-TTF file
 * several fixups

Version 0.3pl1, Folke Ashberg <folke@ashberg.de>
 + support for magic_quotes_gpc set to off

Version 0.3, Folke Ashberg <folke@ashberg.de>
 + Windows compatible
 + various code-fixes


test.php
<?php

include('php-barcode.php');
barcode_save('9783161484100','isbn',2,'jpeg', 'test.jpg');

Tests/Utility/BarcodeTest.php
<?php

namespace Hackzilla\BarcodeBundle\Tests\Utility;

use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;
use Hackzilla\BarcodeBundle\Utility\Barcode;

class BarcodeTest extends WebTestCase
{
    private $_object;

    public function setup()
    {
        $this->_object = new Barcode();
    }

    public function testBarColor()
    {
        $this->_object->setBarColor(array(123,456,789));
        $this->assertEquals(123, $this->_object->barColor(0));
        $this->assertEquals(456, $this->_object->barColor(1));
        $this->assertEquals(789, $this->_object->barColor(2));

        $this->_object->setBarColor(0);
        $this->assertEquals(123, $this->_object->barColor(0));
        $this->assertEquals(456, $this->_object->barColor(1));
        $this->assertEquals(789, $this->_object->barColor(2));
    }

    public function testBgColor()
    {
        $this->_object->setBgColor(array(123,456,789));
        $this->assertEquals(123, $this->_object->bgColor(0));
        $this->assertEquals(456, $this->_object->bgColor(1));
        $this->assertEquals(789, $this->_object->bgColor(2));

        $this->_object->setBgColor(0);
        $this->assertEquals(123, $this->_object->bgColor(0));
        $this->assertEquals(456, $this->_object->bgColor(1));
        $this->assertEquals(789, $this->_object->bgColor(2));
    }

    public function testTextColor()
    {
        $this->_object->setTextColor(array(123,456,789));
        $this->assertEquals(123, $this->_object->textColor(0));
        $this->assertEquals(456, $this->_object->textColor(1));
        $this->assertEquals(789, $this->_object->textColor(2));

        $this->_object->setTextColor(0);
        $this->assertEquals(123, $this->_object->textColor(0));
        $this->assertEquals(456, $this->_object->textColor(1));
        $this->assertEquals(789, $this->_object->textColor(2));
    }

    public function testFontLocation()
    {
        $this->_object->setFontLocation('/path/testing');
        $this->assertEquals('/path/testing', $this->_object->fontLocation());
    }

    public function testGenBarcodeLocation()
    {
        $this->_object->setGenBarcodeLocation('/path/testing');
        $this->assertEquals('/path/testing', $this->_object->genBarcodeLocation());
    }

    public function testEncoding()
    {
        $this->_object->setEncoding(Barcode::encoding_any);
        $this->assertEquals(Barcode::encoding_any, $this->_object->encoding());
    }

    public function testMode()
    {
        $this->_object->setMode(Barcode::mode_png);
        $this->assertEquals(Barcode::mode_png, $this->_object->mode());
    }

    public function testScale()
    {
        $this->_object->setScale(0);
        $this->assertEquals(2, $this->_object->scale());

        $this->_object->setScale(-100);
        $this->assertEquals(2, $this->_object->scale());

        $this->_object->setScale(100);
        $this->assertEquals(100, $this->_object->scale());
    }

    public function testSpace()
    {
        $this->_object->setSpace(array(
            'top' => 12,
            'left' => 34,
            'bottom' => 56,
            'right' => 78,
        ));

        $space = $this->_object->space();
        $this->assertEquals(12, $space['top']);
        $this->assertEquals(34, $space['left']);
        $this->assertEquals(56, $space['bottom']);
        $this->assertEquals(78, $space['right']);
    }

    public function testHeight()
    {
        $this->_object->setHeight(0);
        $this->assertEquals(1, $this->_object->height());

        $this->_object->setHeight(-100);
        $this->assertEquals(1, $this->_object->height());

        $this->_object->setHeight(100);
        $this->assertEquals(100, $this->_object->height());
    }

    public function testChecksum()
    {
        $this->assertEquals(2, $this->_object->generateEanChecksum('500015941539'));
        $this->assertEquals(2, $this->_object->generateEanChecksum('978085934063'));

        $this->assertNotEquals(7, $this->_object->generateEanChecksum('50001594159'));
    }
    
    public function testEncodeEan()
    {
        $bars = $this->_object->encodeEan('978085934063');
        $this->assertEquals('ISBN', $bars['encoding']);
        $this->assertEquals('9a1a1312312111231213132131121a1a1141111323211111414112122a1a', $bars['bars']);
        $this->assertEquals('0:12:9 12:12:7 19:12:8 26:12:0 33:12:8 40:12:5 47:12:9 59:12:3 66:12:4 73:12:0 80:12:6 87:12:3 94:12:2', $bars['text']);
        
        $bars = $this->_object->encodeEan('fail');
        $this->assertEquals('Invalid', \substr($bars['text'], 0, 7));
    }
}

barcode.php
<?php
/*

 * Image-Creator / Sample
 * Part of PHP-Barcode 0.4
 
 * (C) 2001,2002,2003,2004 by Folke Ashberg <folke@ashberg.de>
 
 * The newest version can be found at http://www.ashberg.de/php-barcode
 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 */

require("php-barcode.php");

function getvar($name){
    global $_GET, $_POST;
    if (isset($_GET[$name])) return $_GET[$name];
    else if (isset($_POST[$name])) return $_POST[$name];
    else return false;
}

if (get_magic_quotes_gpc()){
    $code=stripslashes(getvar('code'));
} else {
    $code=getvar('code');
}
if (!$code) $code='123456789012';

barcode_print($code,getvar('encoding'),getvar('scale'),getvar('mode'));

/*
 * call
 * http://........./barcode.php?code=012345678901
 *   or
 * http://........./barcode.php?code=012345678901&encoding=EAN&scale=4&mode=png
 *
 */


encode_bars.php
<?php
/*

 * Built-In Encoders
 * Part of PHP-Barcode 0.4
 
 * (C) 2001,2002,2003,2004,2011 by Folke Ashberg <folke@ashberg.de>
 
 * The newest version can be found at http://www.ashberg.de/php-barcode
 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 */

function barcode_gen_ean_sum($ean){
  $even=true; $esum=0; $osum=0;
  for ($i=strlen($ean)-1;$i>=0;$i--){
	if ($even) $esum+=$ean[$i];	else $osum+=$ean[$i];
	$even=!$even;
  }
  return (10-((3*$esum+$osum)%10))%10;
}

/* barcode_encode_ean(code [, encoding])
 *   encodes $ean with EAN-13 using builtin functions
 *
 *   return:
 *    array[encoding] : the encoding which has been used (EAN-13)
 *    array[bars]     : the bars
 *    array[text]     : text-positioning info
 */
function barcode_encode_ean($ean, $encoding = "EAN-13"){
    $digits=array(3211,2221,2122,1411,1132,1231,1114,1312,1213,3112);
    $mirror=array("000000","001011","001101","001110","010011","011001","011100","010101","010110","011010");
    $guards=array("9a1a","1a1a1","a1a");

    $ean=trim($ean);
    if (preg_match("#[^0-9]#i",$ean)){
	return array("text"=>"Invalid EAN-Code");
    }
    $encoding=strtoupper($encoding);
    if ($encoding=="ISBN"){
	if (!preg_match("#^978#", $ean)) $ean="978".$ean;
    }
    if (preg_match("#^978#", $ean)) $encoding="ISBN";
    if (strlen($ean)<12 || strlen($ean)>13){
	return array("text"=>"Invalid $encoding Code (must have 12/13 numbers)");
    }

    $ean=substr($ean,0,12);
    $eansum=barcode_gen_ean_sum($ean);
    $ean.=$eansum;
    $line=$guards[0];
    for ($i=1;$i<13;$i++){
	$str=$digits[$ean[$i]];
	if ($i<7 && $mirror[$ean[0]][$i-1]==1) $line.=strrev($str); else $line.=$str;
	if ($i==6) $line.=$guards[1];
    }
    $line.=$guards[2];

    /* create text */
    $pos=0;
    $text="";
    for ($a=0;$a<13;$a++){
	if ($a>0) $text.=" ";
	$text.="$pos:12:{$ean[$a]}";
	if ($a==0) $pos+=12;
	else if ($a==6) $pos+=12;
	else $pos+=7;
    }

    return array(
		"encoding" => $encoding,
		"bars" => $line,
		"text" => $text
		);
}


Utility/Barcode.php
<?php

namespace Hackzilla\BarcodeBundle\Utility;

/*
 * (C) 2001,2002,2003,2004,2011 by Folke Ashberg <folke@ashberg.de>
 * (c) 2013 by Daniel Platt <github@ofdan.co.uk>

 * The previous version can be found at http://www.ashberg.de/php-barcode
 * The newest version can be found at http://github.com/hackzilla/barcode-bundle

 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

/**
 * Barcode
 *
 */
class Barcode
{
    private $twig;
    private $barColor;
    private $bgColour;
    private $textColour;
    private $fontLocation;
    private $genbarcodeLocation;

    private $encoding;
    private $mode;
    private $scale;
    private $height;
    private $space;

    const mode_png = 'png';
    const mode_jpeg = 'jpg';
    const mode_gif = 'gif';

    const encoding_any = 'ANY'; // choose best-fit (default)
    const encoding_ean = 'EAN'; // 8 or 13 EAN-Code
    const encoding_upc = 'UPC'; // 12-digit EAN
    const encoding_isbn = 'ISBN'; // isbn numbers (still EAN-13)
    const encoding_39 = '39'; // code39
    const encoding_128 = '128'; // (a,b,c: autoselection)
    const encoding_128c = '128C'; // (compact form for digits)
    const encoding_128b = '128B'; // full printable ascii
    const encoding_i25 = 'I25'; // interleaved 2 of 5 (only digits)
    const encoding_128raw = '128RAW'; // Raw code 128 (by Leonid A. Broukhis)
    const encoding_cbr = 'CBR'; // Codabar (by Leonid A. Broukhis)
    const encoding_msi = 'MSI'; // MSI (by Leonid A. Broukhis)
    const encoding_pls = 'PLS'; // Plessey (by Leonid A. Broukhis)

    public function __construct($container = null)
    {
        if ($container) {
            $this->twig = $container->get('templating');
        }

        $this->setBarColor(array(0, 0, 0));
        $this->setBgColor(array(255, 255, 255));
        $this->setTextColor(array(0, 0, 0));

        $this->setFontLocation(dirname(__FILE__) . "/../Resources/font/FreeSansBold.ttf");

        //$this->setGenbarcodeLocation("c:\winnt\genbarcode.exe");
        $this->setGenbarcodeLocation("/usr/local/bin/genbarcode");

        $this->setHeight(120);

        $this->setEncoding(self::encoding_any);
        $this->setMode(self::mode_png);
        
        $this->setScale(2);
        $this->setSpace();
    }

    public function barColor($id)
    {
        return $this->barColor[$id];
    }

    public function bgColor($id)
    {
        return $this->bgColor[$id];
    }

    public function textColor($id)
    {
        return $this->textColor[$id];
    }

    public function fontLocation()
    {
        return $this->fontLocation;
    }

    public function genbarcodeLocation()
    {
        return $this->genbarcodeLocation;
    }

    public function height()
    {
        return $this->height;
    }

    public function encoding()
    {
        return $this->encoding;
    }

    public function mode()
    {
        return $this->mode;
    }

    public function scale()
    {
        return $this->scale;
    }

    public function space()
    {
        return $this->space;
    }

    public function setBarColor($color)
    {
        if (\is_array($color) && \count($color) == 3) {
            $this->barColor = $color;
        }

        return $this;
    }

    public function setBgColor($color)
    {
        if (\is_array($color) && \count($color) == 3) {
            $this->bgColor = $color;
        }

        return $this;
    }

    public function setTextColor($color)
    {
        if (\is_array($color) && \count($color) == 3) {
            $this->textColor = $color;
        }

        return $this;
    }

    public function setFontLocation($location)
    {
        if (\is_string($location) && \strlen($location) > 0) {
            $this->fontLocation = $location;
        }

        return $this;
    }

    public function setGenbarcodeLocation($location)
    {
        if (\is_string($location) && \strlen($location) > 0) {
            $this->genbarcodeLocation = $location;
        }

        return $this;
    }

    public function setEncoding($encoding)
    {
        $this->encoding = \strtoupper($encoding);

        return $this;
    }

    public function setMode($mode)
    {
        $this->mode = $mode;

        return $this;
    }

    public function setScale($scale)
    {
        if ($scale < 1) {
            $scale = 2;
        }

        $this->scale = $scale;

        return $this;
    }

    public function setSpace($space = false)
    {
        $size = 2 * $this->scale();

        if($space === false || is_array($space)) {
            $this->space = array(
                'top' => $size,
                'left' => $size,
                'bottom' => $size,
                'right' => $size,
            );
        }

        if (is_array($space)) {
            if (isset($space['top']) && is_numeric($space['top'])) {
                $this->space['top'] = $space['top'];
            }
            if (isset($space['left']) && is_numeric($space['left'])) {
                $this->space['left'] = $space['left'];
            }
            if (isset($space['bottom']) && is_numeric($space['bottom'])) {
                $this->space['bottom'] = $space['bottom'];
            }
            if (isset($space['right']) && is_numeric($space['right'])) {
                $this->space['right'] = $space['right'];
            }
        }

        return $this;
    }

    public function setHeight($height)
    {
        if($height < 1) {
            $height = 1;
        }

        $this->height = $height;

        return $this;
    }

    /**
     * Outputs an image using libgd
     * 
     * @param string $code
     */
    public function outputImage($code)
    {
        $im = $this->build($code);

        /* output the image */
        if ($this->mode() == self::mode_jpeg) {
            \header("Content-Type: image/jpeg; name=\"barcode.jpg\"");
            \imagejpeg($im);
        } else if ($this->mode() == self::mode_gif) {
            \header("Content-Type: image/gif; name=\"barcode.gif\"");
            \imagegif($im);
        } else {
            \header("Content-Type: image/png; name=\"barcode.png\"");
            \imagepng($im);
        }
    }

    /**
     *  Returns (!) a barcode as plain-text
     *  ATTENTION: this is very silly!
     * 
     * @param string $code
     * 
     * @return string
     */
    public function outputText($code)
    {
        $bars = $this->encode($code);
        $bars = $bars['bars'];
        
        $width = true;
        $xpos = $heigh2 = 0;
        $bar_line = "";

        for ($i = 0; $i < \strlen($bars); $i++) {
            $val = \strtolower($bars[$i]);
            if ($width) {
                $xpos+=$val;
                $width = false;
                for ($a = 0; $a < $val; $a++) {
                    $bar_line.="-";
                }
                continue;
            }
            if (\preg_match("#[a-z]#", $val)) {
                $val = \ord($val) - \ord('a') + 1;
                $h = $heigh2;
                for ($a = 0; $a < $val; $a++) {
                    $bar_line.="I";
                }
            } else {
                for ($a = 0; $a < $val; $a++) {
                    $bar_line.="#";
                }
            }
            $xpos+=$val;
            $width = true;
        }

        return $bar_line;
    }

    /**
     * HTML-Code for barcode-image using html-code (using a table and with black.png and white.png)
     * 
     * @param string $code
     * 
     * @return string
     */
    public function outputHtml($code)
    {
        $bars = $this->encode($code);
        $bars = $bars['bars'];
        $outBars = array();
     
        $total_y = $this->height();
        $scale = $this->scale();
        $space = $this->space();

        /* generate html-code */
        $height = round($total_y - ($scale * 10));
        $height2 = round($total_y) - $space['bottom'];

        $width = true;
        for ($i = 0; $i < strlen($bars); $i++) {
            $val = strtolower($bars[$i]);
            if ($width) {
                $w = $val * $scale;
                if ($w > 0) {
                    $outBars[] = array(
                        'type' => 'white',
                        'height' => $total_y,
                        'width' => $w,
                    );
                }
                $width = false;
                continue;
            }
            if (preg_match("#[a-z]#", $val)) {
                //hoher strich
                $val = ord($val) - ord('a') + 1;
                $h = $height2;
            } else {
                $h = $height;
            }
            $w = $val * $scale;
            if ($w > 0) {
                $outBars[] = array(
                    'type' => 'black',
                    'height' => $h,
                    'width' => $w,
                );
            }
            $width = true;
        }

        if(\is_object($this->twig)) {
            $out = $this->twig->render('HackzillaBarcodeBundle:Barcode:layout.html.twig', array(
                'height2' => $height2,
                'space_top' => $space['top'],
                'space_bottom' => $space['bottom'],
                'space_left' => $space['left'],
                'space_right' => $space['right'],
                'bars' => $outBars,
            ));
        } else {
            $out = '<p>Twig not enabled in bundle</p>';
        }
        
        return $out;
    }

    /**
     * encodes $code with $encoding using genbarcode
     * 
     * @param string $code
     * 
     * @return array
     *    array[encoding] : the encoding which has been used
     *    array[bars]     : the bars
     *    array[text]     : text-positioning info
     */
    public function encodeGenbarcode($code)
    {
        /* delete EAN-13 checksum */
        if (\preg_match("#^ean$#i", $encoding) && strlen($code) == 13) {
            $code = \substr($code, 0, 12);
        }
        $encoding = \preg_replace("#[|\\\\]#", "_", $this->encoding());
        $code = \preg_replace("#[|\\\\]#", "_", $code);
        $cmd = $this->genbarcodeLocation() . " "
                . \escapeshellarg($code) . " "
                . \escapeshellarg($encoding) . "";
//print "'$cmd'<BR>\n";
        $fp = \popen($cmd, "r");
        if ($fp) {
            $bars = \fgets($fp, 1024);
            $text = \fgets($fp, 1024);
            $encoding = \fgets($fp, 1024);
            \pclose($fp);
        } else {
            return false;
        }
        $ret = array(
            "encoding" => \trim($encoding),
            "bars" => \trim($bars),
            "text" => \trim($text)
        );
        if (!$ret['encoding']) {
            return false;
        }
        if (!$ret['bars']) {
            return false;
        }
        if (!$ret['text']) {
            return false;
        }
        return $ret;
    }

    /**
     * 
     * @param string $code
     * 
     * @return boolean|array
     *    array[encoding] : the encoding which has been used
     *    array[bars]     : the bars
     *    array[text]     : text-positioning info
     */
    public function encode($code)
    {
        if (
                ((preg_match("#^ean$#i", $this->encoding()) && ( strlen($code) == 12 || strlen($code) == 13))) || (($this->encoding()) && (preg_match("#^isbn$#i", $this->encoding())) && (( strlen($code) == 9 || strlen($code) == 10) ||
                (((preg_match("#^978#", $code) && strlen($code) == 12) ||
                (strlen($code) == 13))))) || ((!$this->encoding() || (preg_match("#^ANY$#i", $this->encoding()) )) && (preg_match("#^[0-9]{12,13}$#", $code)))
        ) {
            /* use built-in EAN-Encoder */
            $bars = $this->encodeEan($code);
        } else if (file_exists($this->genbarcodeLocation())) {
            /* use genbarcode */
            $bars = $this->encodeGenbarcode($code);
        } else {
            print "BarcodeBundle needs an external programm for encodings other then EAN/ISBN<BR>\n";
            print "<ul>\n";
            print "<li>download gnu-barcode from <a href=\"http://www.gnu.org/software/barcode/\">www.gnu.org/software/barcode/</a></li>\n";
            print "<li>compile and install them</li>\n";
            print "<li>download genbarcode from <a href=\"http://www.ashberg.de/php-barcode/\">www.ashberg.de/php-barcode/</a></li>\n";
            print "<li>compile and install them</li>\n";
            print "<li>specify path to genbarcode in php-barcode.php</li>\n";
            print "</ul>\n";
            print "<br />\n";
            print "<a href=\"http://www.ashberg.de/php-barcode/\">Folke Ashberg's OpenSource PHP-Barcode</a><br />\n";

            return false;
        }
        return $bars;
    }

    /**
     * encodes and prints a barcode
     * 
     * @param string $code
     */
    public function output($code)
    {
        $mode = $this->mode();

        if (\preg_match("#^(text|txt|plain)$#i", $mode)) {
            print $this->outputText($code);
        } else if (\preg_match("#^(html|htm)$#i", $mode)) {
            print $this->outputHtml($code);
        } else {
            $this->outputImage($code);
        }
    }

    public function build($code, $filename = false)
    {
        $bars = $this->encode($code);

        if (!$bars) {
            return;
        }

        if (!$filename) {
            $filename = $code . '.' . $this->mode();
        }

        $text = $bars['text'];
        $bars = $bars['bars'];
        $total_y = $this->height();
        $scale = $this->scale();
        $space = $this->space();

        /* count total width */
        $xpos = 0;
        $width = true;
        for ($i = 0; $i < \strlen($bars); $i++) {
            $val = \strtolower($bars[$i]);
            if ($width) {
                $xpos+=$val * $scale;
                $width = false;
                continue;
            }
            if (\preg_match("#[a-z]#", $val)) {
                /* tall bar */
                $val = \ord($val) - \ord('a') + 1;
            }
            $xpos+=$val * $scale;
            $width = true;
        }

        /* allocate the image */
        $total_x = $xpos + $space['right'] + $space['right'];
        $xpos = $space['left'];

        $im = \imagecreate($total_x, $total_y);

        /* create two images */
        $col_bg = \ImageColorAllocate($im, $this->bgColor(0), $this->bgColor(1), $this->bgColor(2));
        $col_bar = \ImageColorAllocate($im, $this->barColor(0), $this->barColor(1), $this->barColor(2));
        $col_text = \ImageColorAllocate($im, $this->textColor(0), $this->textColor(1), $this->textColor(2));
        $height = \round($total_y - ($scale * 10));
        $height2 = \round($total_y - $space['bottom']);


        /* paint the bars */
        $width = true;
        for ($i = 0; $i < \strlen($bars); $i++) {
            $val = \strtolower($bars[$i]);
            if ($width) {
                $xpos+=$val * $scale;
                $width = false;
                continue;
            }
            if (\preg_match("#[a-z]#", $val)) {
                /* tall bar */
                $val = \ord($val) - \ord('a') + 1;
                $h = $height2;
            } else {
                $h = $height;
            }
            \imagefilledrectangle($im, $xpos, $space['top'], $xpos + ($val * $scale) - 1, $h, $col_bar);
            $xpos+=$val * $scale;
            $width = true;
        }

        /* write out the text */
        $chars = \explode(" ", $text);
        \reset($chars);
        while (list($n, $v) = each($chars)) {
            if (trim($v)) {
                $inf = explode(":", $v);
                $fontsize = $scale * ($inf[1] / 1.8);
                $fontheight = $total_y - ($fontsize / 2.7) + 2;
                \imagettftext($im, $fontsize, 0, $space['left'] + ($scale * $inf[0]) + 2, $fontheight, $col_text, $this->fontLocation(), $inf[2]);
            }
        }

        return $im;
    }

    /**
     * Save an image of the barcode to $filename or a generated filename
     * 
     * @param string $code
     *
     * @return string $filename
     */
    public function save($code, $filename = false)
    {
        if(!$filename) {
            $filename = tmpfile();
        }

        $im = $this->build($code, $encoding, $scale, $mode, $filename);
        $mode = $this->mode();

        /* output the image */
        if ($mode == self::jpeg) {
            imagejpeg($im, $filename);
        } else if ($mode == self::gif) {
            imagegif($im, $filename);
        } else {
            imagepng($im, $filename);
        }

        return $filename;
    }


    /**
     * Generate the Ean Checksum
     * Pass in first 12 digits
     *
     * @param string $ean (without checksum)
     *
     * @return int
     */
    public function generateEanChecksum($ean)
    {
        $even = true;
        $esum = 0;
        $osum = 0;

        for ($i = strlen($ean) - 1; $i >= 0; $i--) {
            if ($even) {
                $esum+=$ean[$i];
            } else {
                $osum+=$ean[$i];
            }
            $even = !$even;
        }
        return (10 - ((3 * $esum + $osum) % 10)) % 10;
    }

    /**
     * encodes $ean with EAN-13 using builtin functions
     *
     * @param type $ean
     * 
     * @return array
     *    array[encoding] : the encoding which has been used (EAN-13)
     *    array[bars]     : the bars
     *    array[text]     : text-positioning info
     */
    public function encodeEan($ean)
    {
        $digits = array(3211, 2221, 2122, 1411, 1132, 1231, 1114, 1312, 1213, 3112);
        $mirror = array("000000", "001011", "001101", "001110", "010011", "011001", "011100", "010101", "010110", "011010");
        $guards = array("9a1a", "1a1a1", "a1a");

        $ean = \trim($ean);
        $encoding = $this->encoding();

        if (\preg_match("#[^0-9]#i", $ean)) {
            return array("text" => "Invalid EAN-Code");
        }
        if ($encoding == "ISBN") {
            if (!\preg_match("#^978#", $ean))
                $ean = "978" . $ean;
        }
        if (\preg_match("#^978#", $ean)) {
            $encoding = "ISBN";
        }
        if (\strlen($ean) < 12 || \strlen($ean) > 13) {
            return array("text" => "Invalid {$encoding} Code (must have 12/13 numbers)");
        }

        $ean = \substr($ean, 0, 12);
        $eansum = $this->generateEanChecksum($ean);
        $ean .= $eansum;
        $line = $guards[0];

        for ($i = 1; $i < 13; $i++) {
            $str = $digits[$ean[$i]];
            if ($i < 7 && $mirror[$ean[0]][$i - 1] == 1) {
                $line .= strrev($str);
            } else {
                $line .= $str;
            }
            if ($i == 6) {
                $line .= $guards[1];
            }
        }
        $line .= $guards[2];

        /* create text */
        $pos = 0;
        $text = "";
        for ($a = 0; $a < 13; $a++) {
            if ($a > 0) {
                $text.=" ";
            }

            $text.="$pos:12:{$ean[$a]}";
            if ($a == 0) {
                $pos+=12;
            } else if ($a == 6) {
                $pos+=12;
            } else {
                $pos+=7;
            }
        }

        return array(
            "encoding" => $encoding,
            "bars" => $line,
            "text" => $text
        );
    }

}
